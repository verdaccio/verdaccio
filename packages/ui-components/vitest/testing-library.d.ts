import '@testing-library/jest-dom';

declare global {
  namespace jest {
    interface Matchers<R> {
      toBeInTheDocument(): R;
      toHaveTextContent(text: string | RegExp): R;
      toBeVisible(): R;
      toBeDisabled(): R;
      toBeEnabled(): R;
      toHaveAttribute(attr: string, value?: string): R;
      toHaveClass(className: string): R;
      toHaveFocus(): R;
      toBeChecked(): R;
      toBeEmpty(): R;
      toBeEmptyDOMElement(): R;
      toBeInvalid(): R;
      toBeRequired(): R;
      toBeValid(): R;
      toContainElement(element: HTMLElement | null): R;
      toContainHTML(htmlText: string): R;
      toHaveStyle(css: string | object): R;
      toHaveValue(value?: string | string[] | number): R;
    }
  }
}

// Make the file a module
export {};
