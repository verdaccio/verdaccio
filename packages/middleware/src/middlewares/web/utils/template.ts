import buildDebug from 'debug';

import { TemplateUIOptions } from '@verdaccio/types';

import { Manifest, getManifestValue } from './manifest';

const debug = buildDebug('verdaccio:web:render:template');

export type Template = {
  manifest: Manifest;
  options: TemplateUIOptions;
  metaScripts?: string[];
  scriptsBodyAfter?: string[];
  scriptsBodyBefore?: string[];
};

// the outcome of the Webpack Manifest Plugin
export interface WebpackManifest {
  [key: string]: string;
}

export default function renderTemplate(template: Template, manifest: WebpackManifest) {
  debug('template %o', template);
  debug('manifest %o', manifest);

  return `
    <!DOCTYPE html>
      <html lang="en-us">
      <head>
        <meta charset="utf-8">
        <base href="${template?.options.base}">
        <title>${template?.options?.title ?? ''}</title>
        <link rel="icon" href="${template?.options.base}-/static/favicon.ico">
        <meta name="viewport" content="width=device-width, initial-scale=1">
        <script>
            window.__VERDACCIO_BASENAME_UI_OPTIONS=${JSON.stringify(template.options)}
        </script>
        ${template?.metaScripts ? template.metaScripts.join('') : ''}
      </head>
      <body class="body">
        ${template?.scriptsBodyBefore ? template.scriptsBodyBefore.join('') : ''}
        <div id="root"></div>
        ${getManifestValue(template.manifest.js, manifest, template?.options.base)
          .map((item) => `<script defer="defer" src="${item}"></script>`)
          .join(`\n        `)}
        ${template?.scriptsBodyAfter ? template.scriptsBodyAfter.join('') : ''}
      </body>
    </html>
  `;
}
