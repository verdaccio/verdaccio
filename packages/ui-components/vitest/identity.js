/* eslint-disable @typescript-eslint/no-unused-vars */
let idObj;

idObj = new Proxy(
  {},
  {
    get: function getter(target, key) {
      if (key === '__esModule') {
        return false;
      }
      return key;
    },
  }
);

module.exports = idObj;
