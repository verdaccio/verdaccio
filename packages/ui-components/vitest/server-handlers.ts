import fs from 'fs';
import { HttpResponse, http } from 'msw';
import path from 'path';

export const handlers = [
  // Home
  http.get('http://localhost:9000/-/verdaccio/data/packages', () => {
    return HttpResponse.json(Array(400).fill(require('./api/home-packages.json')));
  }),

  // Search
  http.get('http://localhost:9000/-/verdaccio/data/search/*', () => {
    return HttpResponse.json(require('./api/search-verdaccio.json'));
  }),

  // Storybook
  http.get('http://localhost:9000/-/verdaccio/data/sidebar/storybook', ({ params }) => {
    const { v } = params;
    if (v) {
      return HttpResponse.json(require('./api/storybook-v.json'));
    } else {
      return HttpResponse.json(require('./api/storybook-sidebar.json'));
    }
  }),
  http.get('http://localhost:9000/-/verdaccio/data/package/readme/storybook', () => {
    return HttpResponse.text(require('./api/storybook-readme')());
  }),

  // JQuery (with complete Readme)
  http.get('http://localhost:9000/-/verdaccio/data/sidebar/jquery', () => {
    return HttpResponse.json(require('./api/jquery-sidebar.json'));
  }),
  http.get('http://localhost:9000/-/verdaccio/data/package/readme/jquery', () => {
    return HttpResponse.text(require('./api/jquery-readme')());
  }),

  // Sidebar Errors
  http.get('http://localhost:9000/-/verdaccio/data/sidebar/JSONStream', () => {
    return new HttpResponse('unauthorized', { status: 401 });
  }),
  http.get('http://localhost:9000/-/verdaccio/data/sidebar/semver', () => {
    return new HttpResponse('internal server error', { status: 500 });
  }),
  http.get('http://localhost:9000/-/verdaccio/data/sidebar/kleur', () => {
    return new HttpResponse('not found', { status: 404 });
  }),

  // Glob
  http.get('http://localhost:9000/-/verdaccio/data/sidebar/glob', () => {
    return HttpResponse.json(require('./api/glob-sidebar.json'));
  }),
  http.get('http://localhost:9000/-/verdaccio/data/package/readme/glob', () => {
    return HttpResponse.text('foo glob');
  }),

  // Got
  http.get('http://localhost:9000/-/verdaccio/data/sidebar/got', () => {
    return HttpResponse.json(require('./api/got-sidebar.json'));
  }),
  http.get('http://localhost:9000/-/verdaccio/data/package/readme/got', () => {
    return HttpResponse.text('foo got');
  }),

  // Tarball Download
  http.get('http://localhost:9000/verdaccio/-/verdaccio-1.0.0.tgz', () => {
    const fileName = path.resolve(__dirname, './api/verdaccio-1.0.0.tgz');
    const fileContent = fs.readFileSync(fileName);
    return new HttpResponse(fileContent, {
      status: 200,
      headers: { 'Content-Type': 'application/octet-stream' },
    });
  }),

  // Login
  http.post<{ username: string; password: string }, { token: string; username: string }>(
    'http://localhost:9000/-/verdaccio/sec/login',
    async ({ request }) => {
      const body = await request.json();
      if (body.username === 'fail') {
        return new HttpResponse('unauthorized', { status: 401 });
      }

      return HttpResponse.json({ username: body.username, token: 'valid token' });
    }
  ),
];
