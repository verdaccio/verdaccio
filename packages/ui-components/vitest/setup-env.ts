// / <reference path="./testing-library.d.ts" />
import '@testing-library/jest-dom';
import 'whatwg-fetch';

import { server } from './server';

// Configure the test environment URL
Object.defineProperty(window, 'location', {
  writable: true,
  value: new URL('http://localhost:9000/'),
});

beforeAll(() => {
  // Enable debug logging
  // server.events.on('request:start', (req) => {
  //   console.log('Request started:', req.url.href);
  // });
  // server.events.on('request:match', (req) => {
  //   console.log('Request matched:', req.url.href);
  // });
  // server.events.on('request:unhandled', (req) => {
  //   console.warn('Unhandled request:', req.url.href);
  // });
  // server.events.on('request:end', (req) => {
  //   console.log('Request completed:', req.url.href);
  // });
  server.listen({
    onUnhandledRequest: 'warn',
  });
});
afterEach(() => server.resetHandlers());
afterAll(() => {
  server.close();
});
