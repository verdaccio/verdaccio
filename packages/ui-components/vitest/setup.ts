// / <reference path="./testing-library.d.ts" />

/**
 * Setup configuration for Jest
 * This file includes global settings for the JEST environment.
 */
import '@testing-library/jest-dom';
import 'mutationobserver-shim';
import { vi } from 'vitest';
import { Headers, Request, Response, fetch } from 'whatwg-fetch';

// Override the global fetch and related APIs
global.fetch = fetch;
global.Headers = Headers;
global.Request = Request;
global.Response = Response;

// @ts-ignore : Property '__VERDACCIO_BASENAME_UI_OPTIONS' does not exist on type 'Global'.
global.__VERDACCIO_BASENAME_UI_OPTIONS = {
  base: 'http://localhost:9000/',
  protocol: 'http',
  host: 'localhost',
  primaryColor: '#4b5e40',
  url_prefix: '',
  darkMode: false,
  language: 'en-US',
  uri: 'http://localhost:9000/',
  pkgManagers: ['pnpm', 'yarn', 'npm'],
  title: 'Verdaccio Dev UI',
  scope: '',
  version: 'v1.0.0',
};

// mocking few DOM methods
// @ts-ignore : Property 'document' does not exist on type 'Global'.
if (global.document) {
  // @ts-ignore : Type 'Mock<{ selectNodeContents: () => void; }, []>' is not assignable to type '() => Range'.
  document.createRange = vi.fn((): void => ({
    selectNodeContents: (): void => {},
  }));
  document.execCommand = vi.fn();
}
