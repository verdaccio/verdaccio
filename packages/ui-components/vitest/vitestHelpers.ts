import { vi } from 'vitest';

export const mockCreateObjectURL = vi.fn();
export const mockRevokeObjectURL = vi.fn();

let originalUserAgent: string;

/**
 * Sets up mocks for URL and navigator.userAgent to handle download functionality in tests
 */
export function setupDownloadMocks(): void {
  originalUserAgent = window.navigator.userAgent;

  URL.createObjectURL = mockCreateObjectURL;
  URL.revokeObjectURL = mockRevokeObjectURL;

  Object.defineProperty(window.navigator, 'userAgent', {
    value: 'jsdom',
    configurable: true,
  });
}

/**
 * Cleans up mocks for URL and navigator.userAgent
 */
export function cleanupDownloadMocks(): void {
  mockCreateObjectURL.mockReset();
  mockRevokeObjectURL.mockReset();

  Object.defineProperty(window.navigator, 'userAgent', {
    value: originalUserAgent,
    configurable: true,
  });
}
