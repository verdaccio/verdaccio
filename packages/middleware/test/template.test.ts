import { describe, expect, test } from 'vitest';

import template from '../src/middlewares/web/utils/template';

const manifest = require('./partials/manifest/manifest.json');

// manifest expected to have leading slash
// see packages\middleware\test\partials\manifest\manifest.json
const exampleManifest = {
  css: ['main.css'],
  js: ['runtime.js', 'main.js'],
  ico: 'favicon.ico',
};

// "base" is expected to be result of getPublicUrl
// i.e. it must be valid URL with trailing slash
describe('template', () => {
  test('custom render', () => {
    expect(
      template({ options: { base: 'http://domain.com/' }, manifest: exampleManifest }, manifest)
    ).toMatchSnapshot();
  });

  test('custom title', () => {
    expect(
      template(
        { options: { base: 'http://domain.com/', title: 'foo title' }, manifest: exampleManifest },
        manifest
      )
    ).toMatchSnapshot();
  });

  test('meta scripts', () => {
    expect(
      template(
        {
          options: { base: 'http://domain.com/' },
          metaScripts: [`<style>.someclass{font-size:10px;}</style>`],
          manifest: exampleManifest,
        },
        manifest
      )
    ).toMatchSnapshot();
  });

  test('custom body after', () => {
    expect(
      template(
        {
          options: { base: 'http://domain.com/' },
          scriptsBodyAfter: [`<script src="foo"/>`],
          manifest: exampleManifest,
        },
        manifest
      )
    ).toMatchSnapshot();
  });

  test('custom body before', () => {
    expect(
      template(
        {
          options: { base: 'http://domain.com/' },
          scriptsBodyBefore: [`<script src="fooBefore"/>`, `<script src="barBefore"/>`],
          manifest: exampleManifest,
        },
        manifest
      )
    ).toMatchSnapshot();
  });
});
