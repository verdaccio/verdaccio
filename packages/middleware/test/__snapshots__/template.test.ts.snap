// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`template > custom body after 1`] = `
"
    <!DOCTYPE html>
      <html lang="en-us">
      <head>
        <meta charset="utf-8">
        <base href="http://domain.com/">
        <title></title>
        <link rel="icon" href="http://domain.com/-/static/favicon.ico">
        <meta name="viewport" content="width=device-width, initial-scale=1">
        <script>
            window.__VERDACCIO_BASENAME_UI_OPTIONS={"base":"http://domain.com/"}
        </script>
        
      </head>
      <body class="body">
        
        <div id="root"></div>
        <script defer="defer" src="http://domain.com/-/static/runtime.6126058572f989c948b1.js"></script>
        <script defer="defer" src="http://domain.com/-/static/main.6126058572f989c948b1.js"></script>
        <script src="foo"/>
      </body>
    </html>
  "
`;

exports[`template > custom body before 1`] = `
"
    <!DOCTYPE html>
      <html lang="en-us">
      <head>
        <meta charset="utf-8">
        <base href="http://domain.com/">
        <title></title>
        <link rel="icon" href="http://domain.com/-/static/favicon.ico">
        <meta name="viewport" content="width=device-width, initial-scale=1">
        <script>
            window.__VERDACCIO_BASENAME_UI_OPTIONS={"base":"http://domain.com/"}
        </script>
        
      </head>
      <body class="body">
        <script src="fooBefore"/><script src="barBefore"/>
        <div id="root"></div>
        <script defer="defer" src="http://domain.com/-/static/runtime.6126058572f989c948b1.js"></script>
        <script defer="defer" src="http://domain.com/-/static/main.6126058572f989c948b1.js"></script>
        
      </body>
    </html>
  "
`;

exports[`template > custom render 1`] = `
"
    <!DOCTYPE html>
      <html lang="en-us">
      <head>
        <meta charset="utf-8">
        <base href="http://domain.com/">
        <title></title>
        <link rel="icon" href="http://domain.com/-/static/favicon.ico">
        <meta name="viewport" content="width=device-width, initial-scale=1">
        <script>
            window.__VERDACCIO_BASENAME_UI_OPTIONS={"base":"http://domain.com/"}
        </script>
        
      </head>
      <body class="body">
        
        <div id="root"></div>
        <script defer="defer" src="http://domain.com/-/static/runtime.6126058572f989c948b1.js"></script>
        <script defer="defer" src="http://domain.com/-/static/main.6126058572f989c948b1.js"></script>
        
      </body>
    </html>
  "
`;

exports[`template > custom title 1`] = `
"
    <!DOCTYPE html>
      <html lang="en-us">
      <head>
        <meta charset="utf-8">
        <base href="http://domain.com/">
        <title>foo title</title>
        <link rel="icon" href="http://domain.com/-/static/favicon.ico">
        <meta name="viewport" content="width=device-width, initial-scale=1">
        <script>
            window.__VERDACCIO_BASENAME_UI_OPTIONS={"base":"http://domain.com/","title":"foo title"}
        </script>
        
      </head>
      <body class="body">
        
        <div id="root"></div>
        <script defer="defer" src="http://domain.com/-/static/runtime.6126058572f989c948b1.js"></script>
        <script defer="defer" src="http://domain.com/-/static/main.6126058572f989c948b1.js"></script>
        
      </body>
    </html>
  "
`;

exports[`template > meta scripts 1`] = `
"
    <!DOCTYPE html>
      <html lang="en-us">
      <head>
        <meta charset="utf-8">
        <base href="http://domain.com/">
        <title></title>
        <link rel="icon" href="http://domain.com/-/static/favicon.ico">
        <meta name="viewport" content="width=device-width, initial-scale=1">
        <script>
            window.__VERDACCIO_BASENAME_UI_OPTIONS={"base":"http://domain.com/"}
        </script>
        <style>.someclass{font-size:10px;}</style>
      </head>
      <body class="body">
        
        <div id="root"></div>
        <script defer="defer" src="http://domain.com/-/static/runtime.6126058572f989c948b1.js"></script>
        <script defer="defer" src="http://domain.com/-/static/main.6126058572f989c948b1.js"></script>
        
      </body>
    </html>
  "
`;
